File: apps/mobile/lib/features/visitor/domain/visitor.dart
enum VisitorStatus { pending, checkedIn, checkedOut, cancelled }

VisitorStatus _statusFromApi(Object? value) {
  switch (value) {
    case 'CHECKED_IN':
      return VisitorStatus.checkedIn;
    case 'CHECKED_OUT':
      return VisitorStatus.checkedOut;
    case 'CANCELLED':
      return VisitorStatus.cancelled;
    default:
      return VisitorStatus.pending;
  }
}

class Visitor {
  const Visitor({
    required this.id,
    required this.guestName,
    required this.guestPhone,
    required this.visitDate,
    required this.expectedTime,
    required this.vehicleInfo,
    required this.plate,
    required this.purpose,
    required this.notes,
    required this.status,
    required this.isWalkIn,
    required this.houseCode,
    required this.householdDisplayName,
    required this.checkedInAt,
    required this.checkedOutAt,
    required this.createdAt,
    this.qrToken,
  });

  final String id;
  final String guestName;
  final String? guestPhone;

  /// YYYY-MM-DD, matching the API's date-only contract.
  final String visitDate;

  /// HH:mm local wall-clock time, matching the API's time-only contract.
  final String? expectedTime;
  final String? vehicleInfo;
  final String? plate;
  final String? purpose;
  final String? notes;
  final VisitorStatus status;
  final bool isWalkIn;
  final String houseCode;
  final String householdDisplayName;
  final DateTime? checkedInAt;
  final DateTime? checkedOutAt;
  final DateTime createdAt;

  /// Only present when the API includes it — residents (who lack
  /// `visitor.checkin`) get this back on both create and list responses so
  /// they can show the QR code to security; security staff never receive it.
  final String? qrToken;

  factory Visitor.fromJson(Map<String, dynamic> json) {
    final checkedInAt = json['checkedInAt'];
    final checkedOutAt = json['checkedOutAt'];
    return Visitor(
      id: json['id'] as String,
      guestName: json['guestName'] as String,
      guestPhone: json['guestPhone'] as String?,
      visitDate: json['visitDate'] as String,
      expectedTime: json['expectedTime'] as String?,
      vehicleInfo: json['vehicleInfo'] as String?,
      plate: json['plate'] as String?,
      purpose: json['purpose'] as String?,
      notes: json['notes'] as String?,
      status: _statusFromApi(json['status']),
      isWalkIn: json['isWalkIn'] as bool? ?? false,
      houseCode: json['houseCode'] as String,
      householdDisplayName: json['householdDisplayName'] as String,
      checkedInAt: checkedInAt is String ? DateTime.parse(checkedInAt) : null,
      checkedOutAt:
          checkedOutAt is String ? DateTime.parse(checkedOutAt) : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
      qrToken: json['qrToken'] as String?,
    );
  }
}

File: packages/contracts/src/visitor.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const visitorStatusSchema = z.enum([
  "PENDING",
  "CHECKED_IN",
  "CHECKED_OUT",
  "CANCELLED",
]);

export type VisitorStatus = z.infer<typeof visitorStatusSchema>;

export const createVisitorInputSchema = z
  .object({
    guestName: z.string().trim().min(2).max(160),
    guestPhone: z.string().trim().min(6).max(24).optional(),
    visitDate: z
      .string()
      .regex(/^\d{4}-\d{2}-\d{2}$/, "Tanggal harus menggunakan format YYYY-MM-DD."),
    expectedTime: z
      .string()
      .regex(/^(?:[01]\d|2[0-3]):[0-5]\d$/, "Waktu harus menggunakan format HH:mm.")
      .optional(),
    vehicleInfo: z.string().trim().max(160).optional(),
    plate: z.string().trim().max(20).optional(),
    purpose: z.string().trim().max(200).optional(),
    notes: z.string().trim().max(1000).optional(),
  })
  .strict();

export type CreateVisitorInput = z.infer<typeof createVisitorInputSchema>;

export const createWalkInVisitorInputSchema = z
  .object({
    houseCode: z.string().trim().min(1).max(24),
    guestName: z.string().trim().min(2).max(160),
    guestPhone: z.string().trim().min(6).max(24).optional(),
    vehicleInfo: z.string().trim().max(160).optional(),
    plate: z.string().trim().max(20).optional(),
    purpose: z.string().trim().max(200).optional(),
  })
  .strict();

export type CreateWalkInVisitorInput = z.infer<typeof createWalkInVisitorInputSchema>;

export const visitorSchema = z.object({
  id: z.string().uuid(),
  guestName: z.string().min(1),
  guestPhone: z.string().nullable(),
  visitDate: z.string(),
  expectedTime: z.string().nullable(),
  vehicleInfo: z.string().nullable(),
  plate: z.string().nullable(),
  purpose: z.string().nullable(),
  notes: z.string().nullable(),
  status: visitorStatusSchema,
  isWalkIn: z.boolean(),
  houseCode: z.string().min(1),
  householdDisplayName: z.string().min(1),
  checkedInAt: z.string().datetime({ offset: true }).nullable(),
  checkedOutAt: z.string().datetime({ offset: true }).nullable(),
  createdAt: z.string().datetime({ offset: true }),
});

export type Visitor = z.infer<typeof visitorSchema>;

export const visitorWithQrSchema = visitorSchema.extend({
  qrToken: z.string().min(1),
});

export type VisitorWithQr = z.infer<typeof visitorWithQrSchema>;

export const visitorListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(visitorSchema) }),
);

export const visitorMutationResponseSchema = dataEnvelopeSchema(
  z.object({ visitor: visitorWithQrSchema }),
);

export const visitorLookupResponseSchema = dataEnvelopeSchema(
  z.object({ visitor: visitorSchema.nullable() }),
);

export type VisitorListResponse = z.infer<typeof visitorListResponseSchema>;
export type VisitorMutationResponse = z.infer<typeof visitorMutationResponseSchema>;
export type VisitorLookupResponse = z.infer<typeof visitorLookupResponseSchema>;
